`kafka`: fix `alter-config --delete` for some "sticky" properties

WillemKauf opened the pull request:
There was buggy behavior for "sticky"[^1] properties and empty `std::optional<>`s created as a result of `rpk topic alter-config [TOPIC] --delete [PROPERTY]`.

Currently, `redpanda.remote.read` and `redpanda.remote.write` exhibit unique behavior in terms of not falling back to the configured cluster defaults for `cloud_storage_enable_remote_read` and `cloud_storage_enable_remote_write`.
    
We had a bug, triggered by this sequence of events:

1. `cloud_storage_enable_remote_read=false` at cluster level.
2. `rpk topic create t`
3. `rpk topic describe t` (correctly) describes `redpanda.remote.read=false`
4. `rpk cluster config set cloud_storage_enable_remote_read=true`
5. `rpk topic alter-config t --delete redpanda.remote.read`
6. `rpk topic describe t` describes `redpanda.remote.read=true`, but the underlying configuration option is actually `std::nullopt`

We now have `rpk topic describe` printing the cluster default and indicating topic `t` has remote read permissions, but this is not actually the case as far as the topic is concerned (topic property is `std::nullopt`, once again, we don't fall back on cluster default).

Fix the behaviour here by removing code in `incremental_update()` that set the shadow indexing optional to `std::nullopt` when the mode is `shadow_indexing_mode::disabled`. Also, when an `alter-config --delete` request is issued, fall back to the cluster default when setting the new `shadow_indexing_mode`.

This still results in somewhat unique behaviour for these topic properties as opposed to others in `redpanda`, as these properties will take on the value of the cluster default _**at time of call to**_ `alter-config --delete`, but future changes to the cluster default will not affect their values. This is by design, once again according to reference[^1]. 

[^1]: https://github.com/redpanda-data/redpanda/issues/7451

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [X] none - papercut/not impactful enough to backport
- [ ] v24.2.x
- [ ] v24.1.x
- [ ] v23.3.x

## Release Notes

### Bug Fixes

* Fixes a bug that would result in `rpk topic alter-config [TOPIC] --delete [PROPERTY]` causing `redpanda.remote.read` and `redpanda.remote.write` to become unset, instead of referring back to the cluster default.

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/54101#0191c844-0dbd-4c3b-8e6c-1f1355693db0:
```
"rptest.tests.alter_topic_configuration_test.ShadowIndexingGlobalConfig.test_overrides_remove"
```
<!--12b5afdc-383f-48e5-bbe3-d058226e8223-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/54101#0191c846-fb44-4d22-ac35-6d56a4855d3c
<!--9b419e09-72e7-479e-bdcb-ef21ce83b254-->

pgellert reviewed (changes requested):
As discussed offline, we should fix this by changing the behaviour of `rpk topic alter-config t --delete redpanda.remote.read` to reset the topic config to the current cluster-level default instead of (or in addition to) changing the behaviour of describe configs.

WillemKauf commented:
Changed logic of PR to correct expected behavior:
* `alter-config t --delete redpanda.remote.read/write` will now cause these properties to become set to the cluster default _**at time of call**_ from `redpanda_cloud_storage_enable_remote_read/write`. However, they will **_NOT_** follow future updates to the cluster default.
* Removed code that set `shadow_indexing_mode` property to `std::nullopt` when it was equal to `shadow_indexing_mode::disabled`.

WillemKauf commented:
Force push to:
* Correct a small line of Python code in [alter_topic_configuration_test.py](https://github.com/redpanda-data/redpanda/compare/f80fec5f81795ea2dbd30ef16953e5d57b8c0d49..190134b0d8d93cbc0727045ee508f14320ffa02c#diff-0879970555c08129c60c644a81e0f1a62799e61d8e9f026089502075e5db4235)

pgellert commented on src/v/kafka/server/handlers/incremental_alter_configs.cc:
```diff
@@ -49,7 +49,22 @@ static void parse_and_set_shadow_indexing_mode(
     case config_resource_operation::remove:
         simode.op = cluster::incremental_update_operation::remove;
-        simode.value = model::negate_shadow_indexing_flag(enabled_value);
+        switch (enabled_value) {
```
This works when only one of `redpanda.remote.read` or `redpanda.remote.write` is specified in the request. That is not always the case though, which leads to unexpected behaviour such as this example:
```
% rpk topic create t
% rpk cluster config set cloud_storage_enable_remote_write=true
% rpk topic alter-config t --delete redpanda.remote.write --delete redpanda.remote.read
% rpk topic describe t
CONFIGS
=======
KEY                                   VALUE       SOURCE
redpanda.remote.read                  false       DEFAULT_CONFIG
redpanda.remote.write                 false       DYNAMIC_TOPIC_CONFIG

% rpk topic alter-config t --delete redpanda.remote.write
% rpk topic alter-config t --delete redpanda.remote.read
% rpk topic describe t
CONFIGS
=======
KEY                                   VALUE       SOURCE
redpanda.remote.read                  false       DEFAULT_CONFIG
redpanda.remote.write                 true        DEFAULT_CONFIG
```

Today I learned that rpk works around redpanda's inability to change `redpanda.remote.read` and `redpanda.remote.write` at the same time and issues separate `IncrementalAlterConfigs` requests separately. But it only does that when both are `--set` and doesn't have this workaround extended to `--delete`. Ref: https://github.com/redpanda-data/redpanda/blob/74223be8ce5ee37c165bcfcb0187018cdb6653ed/src/go/rpk/pkg/cli/topic/config.go#L75-L78

I think the *right* thing to do would be to deprecate `topic_properties::shadow_indexing` and break it out into two separate fields `remote_read` and `remote_write` to allow changing them independently. But that would be a sizable chunk of work (perhaps something to discuss with the storage team). Without that, I think it would be best to (1) extend the rpk workaround to delete's as well and (2) error out when both properties are specified to let third-party tools know that we can't do that change instead of silently doing the wrong update. I think we would need to error out in alter_configs as well, that seems to have the same issue in some edge cases. @Lazin what do you think?

WillemKauf replied:
Yeah, after running into this bug more and more in adjacent work I think we need to deprecate the `shadow_indexing` property as it exists today and Do The Right (and Simple) thing of keeping `remote.read` and `remote.write` permissions in separate `bool`s.

[I've made a JIRA ticket for this](https://redpandadata.atlassian.net/browse/CORE-7743), but it's unlikely I will get to it within the near future.

WillemKauf replied:
Closing this PR in favour of https://github.com/redpanda-data/redpanda/pull/23545

WillemKauf commented:
Closing this PR, as it has been superseded by https://github.com/redpanda-data/redpanda/pull/23545.
